test/e2e/storage/volume_io.go

```diff
@@ -61,6 +61,134 @@ var md5hashes = map[int64]string{
 	fileSizeLarge:  "8d763edc71bd16217664793b5a15e403",
 }
 
+type volumeIOTestSuite struct {
+	testPatterns map[string]testPattern
+}
+
+func initVolumeIOTestSuite() testSuite {
+	return &volumeIOTestSuite{
+		testPatterns: map[string]testPattern{
+			"Volume": {
+				testVolType: inlineVolume,
+			},
+			"Static PV": {
+				testVolType: staticPV,
+			},
+			"Dynamic PV": {
+				testVolType: dynamicPV,
+			},
+		},
+	}
+}
+
+func (t *volumeIOTestSuite) getTestPatterns() map[string]testPattern {
+	return t.testPatterns
+}
+
+func (t *volumeIOTestSuite) skipUnsupportedTest(testPattern testPattern, driver testDriver) {
+}
+
+func (t *volumeIOTestSuite) initTestResource() testResource {
+	return &genericVolumeTestResource{}
+}
+
+func (t *volumeIOTestSuite) createTestInput(
+	testPattern testPattern,
+	testResource testResource,
+) testInput {
+	if r, ok := testResource.(*genericVolumeTestResource); ok {
+		driver := r.driver
+		driverInfo := driver.getDriverInfo()
+		f := driverInfo.f
+		fileSizes := createFileSizes(driverInfo.maxFileSize)
+		volSource := r.volSource
+
+		if volSource == nil {
+			framework.Skipf("Driver %q does not define volumeSource - skipping", driverInfo.name)
+		}
+
+		var fsGroup int64
+		if driverInfo.isFsGroupSupported {
+			fsGroup = int64(1234)
+		}
+
+		return volumeIOTestInput{
+			f:         f,
+			name:      driverInfo.name,
+			config:    driverInfo.config,
+			volSource: *volSource,
+			testFile:  fmt.Sprintf("%s_io_test_%s", driverInfo.name, f.Namespace.Name),
+			podSec: v1.PodSecurityContext{
+				FSGroup: &fsGroup,
+			},
+			fileSizes: fileSizes,
+		}
+	}
+
+	framework.Failf("Fail to convert testResource to genericVolumeTestResource")
+	return nil
+}
+
+func (t *volumeIOTestSuite) getTestFunc(testPattern testPattern, driver testDriver) func(*testInput) {
+	return testVolumeIOWithTestInput
+}
+
+func testVolumeIOWithTestInput(testInput *testInput) {
+	Context("-", func() {
```

Comment:
It starts with:
```
[sig-storage] In-tree Volumes Driver: NFS, Test: volumeIO, TestPattern: Pre-provisioned PV - 
  should write files of various sizes, verify size, validate content [Slow]
```

And ends with:
```
[sig-storage] In-tree Volumes
/root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/framework.go:22
  Driver: NFS, Test: volumeIO, TestPattern: Pre-provisioned PV
  /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/in_tree_volumes.go:92
    -
    /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/volume_io.go:138
      should write files of various sizes, verify size, validate content [Slow]
      /root/mkimuram/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/volume_io.go:171
------------------------------
```

Reply:
To make it easier to filter, I think we should have driver in brackets, like `[Driver: nfs]`.

Also, instead of context here being basically empty, can we put the test suite name and test pattern here instead?  That way someone could import this file as a library and have their own runner logic.

Reply:
Right now, most of our volume tests follow the format "<feature> <plugin> <detail>".  Does it make sense to continue that, or do you prefer having plugin first?

Reply:
Thank you for your suggestion.

```Context``` here is one of the implementations that I'm not satisfied and this is a tricky part to tame ginkgo to work as we expect. I've spent much time to improve it these days, however I couldn't find a good way.

I applyed pattern 2 of "Global Shared Behaviors" described in https://onsi.github.io/ginkgo/#global-shared-behaviors, it requires all the inputs for ```It()``` to be in one struct. Also, to make various ```testSuite```s share the same logic inside the same loop, I made ```testInput``` interface and use type assertion here to convert each ```testInput``` for each ```testSuite```. 

The problem here is that first string argument of ```Context()``` seems to be evaluated, before ```testInput``` is decided, so if we add a code to make a string depending on ```testInput```, it fails with nil pointer dereference. (Also, it is not allowed to pass another arguments to ```test*WithTestInput()```, because of the above reason.)

I can't find a good documentation about it, but I guess that this is due to ginkgo's multistep execution, explained in https://github.com/onsi/ginkgo/issues/462#issuecomment-384633172. From this explanation, we can guess that context string needs to be decided in "Parse stage" to be properly filtered in "Run stage".
(By the same reason, it seems that type assertion needs to be done inside ```BeforeEach```, or in "Run stage", after ```testInput``` is decided.)

Therefore, I think we can put ```testSuite``` dependent string here, because ```testInput``` and ```testSuite``` have 1:1 relationship now. However, we won't be able to put ```testDriver``` or ```testPattern``` dependent string, which will need to be passed as a part of ```testInput```, here.

So, the choices will be:

1. Remove ugly "-" and output ```<plugin> <detail>```, then ```<feature>```
```
[sig-storage] In-tree Volumes
  [Driver: NFS][TestPattern: Pre-provisioned PV]
    [TestSuite: volumeIO]
      should write files of various sizes, verify size, validate content [Slow]
```
2. Prioritize the order of output and keep ugly "-" (We can replace "-" with any static string, though.)
```
[sig-storage] In-tree Volumes
  [TestSuite: volumeIO][Driver: NFS][TestPattern: Pre-provisioned PV]
    -
      should write files of various sizes, verify size, validate content [Slow]
```
3. Continue to search for a better way

Reply:
Thanks for the detailed explanation!

I think the issue like you said, is that `testInput` is evaluated at runtime.  What if you create a different input, like `testMetadata`, which only contains suite name, driver name, and pattern name, and is initialized during Parsing?

Reply:
Thank you for your suggestion. I tried a similar implementation, but it was not "initialized during Parsing".
I will try this way.